compiler/damlc/tests/src/DamlcTest.hs

```diff
@@ -26,7 +26,113 @@ main = do
+      callProcessSilent damlc ["build", "--project-root", projDir]
+      let dar = projDir </> ".daml/dist/good-0.0.1.dar"
+      step "validate"
+      (exitCode, stdout, stderr) <- readProcessWithExitCode damlc ["validate-dar", dar] ""
+      exitCode @?= ExitSuccess
+      assertInfixOf "DAR is valid" stdout
+      stderr @?= ""
+
+  , testCaseSteps "Bad, DAR contains a bad DALF" $ \step -> withTempDir $ \projDir -> do
+      step "prepare"
+      writeFileUTF8 (projDir </> "daml.yaml") $ unlines
+        [ "sdk-version: " <> sdkVersion
+        , "name: good"
+        , "version: 0.0.1"
+        , "source: ."
+        , "dependencies: [daml-prim, daml-stdlib]"
+        ]
+      writeFileUTF8 (projDir </> "Good.daml") $ unlines
+        [ "daml 1.2 module Good where"
+        , "good = 1"
+        ]
+      step "build"
+      callProcessSilent damlc ["build", "--project-root", projDir]
+      let origDar = projDir </> ".daml/dist/good-0.0.1.dar"
+      let tmpDirForUnzipping = projDir </> "temp-dir-for-unzipping"
+      let hackedDar = projDir </> "hacked.dar"
+      step "unzip"
+      callProcess "/usr/bin/unzip" ["-q", origDar, "-d", tmpDirForUnzipping]
+      step "add bad dalf"
+      writeFileUTF8 (tmpDirForUnzipping </> "rubbish.dalf") $ unlines
+        [ "I am not a DALF file."
+        , "I shouldn't be in the archive with a .dalf extension." ]
+      step "re-zip"
+      callProcess "/usr/bin/zip" ["-q", "-r", hackedDar, tmpDirForUnzipping]
+      step "validate"
+      (exitCode, stdout, stderr) <- readProcessWithExitCode damlc ["validate-dar", hackedDar] ""
+      exitCode @?= ExitFailure 1
+      assertInfixOf "Invalid DAR" stderr
+      assertInfixOf "DALF entry cannot be decoded" stderr
+      stdout @?= ""
+
+  , testCaseSteps "Bad, unclosed" $ \step -> withTempDir $ \projDir -> do
+      step "prepare"
+      writeFileUTF8 (projDir </> "daml.yaml") $ unlines
+        [ "sdk-version: " <> sdkVersion
+        , "name: good"
+        , "version: 0.0.1"
+        , "source: ."
+        , "dependencies: [daml-prim, daml-stdlib]"
+        ]
+      writeFileUTF8 (projDir </> "Good.daml") $ unlines
+        [ "daml 1.2 module Good where"
+        , "good = 1"
+        ]
+      step "build"
+      callProcessSilent damlc ["build", "--project-root", projDir]
+      let origDar = projDir </> ".daml/dist/good-0.0.1.dar"
+      let tmpDirForUnzipping = projDir </> "temp-dir-for-unzipping"
+      let hackedDar = projDir </> "hacked.dar"
+      step "unzip"
+      callProcess "/usr/bin/unzip" ["-q", origDar, "-d", tmpDirForUnzipping]
+      step "find"
+      (_,findOut,_) <- readProcessWithExitCode"/usr/bin/find" [tmpDirForUnzipping] ""
+      step "select file"
+      [selectedFile] <- pure [ filename | filename <- lines findOut, "Date-Types" `isInfixOf` filename ]
+      step "remove file"
+      callProcess "/bin/rm" [selectedFile]
+      step "re-zip"
+      callProcess "/usr/bin/zip" ["-q", "-r", hackedDar, tmpDirForUnzipping]
```

Comment:
Same issue here, this will not work on Windows (and `/usr/bin` doesn’t even work on Linux on things like nixos). You need to get this from `dev-env` by adding `@zip_dev_env///:zip` to `data` of this test and then find it using `locate-runfiles` or alternatively you can just use one of the Haskell zip libs e.g. `zip`.

Reply:
Switch to using `"zip-archive"` haskell library - much simpler!